### PR TITLE
Precache livetv program images

### PR DESCRIPTION
**Pull request by crobibero**
Should fix https://github.com/jellyfin/jellyfin/issues/8835

### CODE REVIEW

**Review by Shadowghost — APPROVED** [review]

File: MediaBrowser.Providers/Manager/ProviderManager.cs [hunk 1]
```diff
@@ -150,52 +162,86 @@ public Task<ItemUpdateType> RefreshSingleItem(BaseItem item, MetadataRefreshOpti
...
+                    {
+                        await SaveImage(
+                            item,
+                            cacheStream,
+                            cachedValue.Value.ContentType,
+                            type,
+                            imageIndex,
+                            cancellationToken).ConfigureAwait(false);
+                    }
+                }
```
Comment (Bond-009) [inline MediaBrowser.Providers/Manager/ProviderManager.cs:182]: Shouldn't we do an early return here if we find it in the cache?

Comment (crobibero) [reply MediaBrowser.Providers/Manager/ProviderManager.cs:183]: I swear I had the return there..

File: MediaBrowser.Providers/Manager/ProviderManager.cs [hunk 2]
```diff
@@ -150,52 +162,87 @@ public Task<ItemUpdateType> RefreshSingleItem(BaseItem item, MetadataRefreshOpti
...
-            var contentType = response.Content.Headers.ContentType?.MediaType;
+                var httpClient = _httpClientFactory.CreateClient(NamedClient.Default);
+                using var response = await httpClient.GetAsync(url, cancellationToken).ConfigureAwait(false);
 
-            // Workaround for tvheadend channel icons
-            // TODO: Isolate this hack into the tvh plugin
-            if (string.IsNullOrEmpty(contentType))
-            {
-                if (url.Contains("/imagecache/", StringComparison.OrdinalIgnoreCase))
+                if (response.StatusCode != HttpStatusCode.OK)
```
Comment (JPVenson) [inline MediaBrowser.Providers/Manager/ProviderManager.cs:188]:
```suggestion
                response.EnsureSuccessStatusCode()
```

File: MediaBrowser.Providers/Manager/ProviderManager.cs [hunk 3]
```diff
@@ -150,52 +162,84 @@ public Task<ItemUpdateType> RefreshSingleItem(BaseItem item, MetadataRefreshOpti
...
-                    cancellationToken).ConfigureAwait(false);
+                var stream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
+                await using (stream.ConfigureAwait(false))
+                {
+                    var memoryStream = new MemoryStream();
+                    await using (memoryStream.ConfigureAwait(false))
+                    {
+                        await stream.CopyToAsync(memoryStream, cancellationToken).ConfigureAwait(false);
+                        memoryStream.Seek(0, SeekOrigin.Begin);
+                        _memoryCache.Set(url, (contentType, memoryStream.GetBuffer()), TimeSpan.FromSeconds(10));
```
Comment (Bond-009) [inline MediaBrowser.Providers/Manager/ProviderManager.cs:232]: `MemoryStream.GetBuffer()` returns the underlying buffer, it's length may not match the actual content
I'd use `HttpContent.ReadAsByteArrayAsync` and construct a `MemoryStream` from that for use by `SaveImage`

**Review by Bond-009 — CHANGES_REQUESTED** [review]

**Review by Bond-009 — APPROVED** [review]

**Review by JPVenson — APPROVED** [review]

**Pull request merged**